File: src/python/oop/ShippingLocker/locker/price_calculator.py
from datetime import datetime, timedelta

from .locker import *

class PriceCalculator:
    def calculate_price(self, locker: Locker) -> float:
        policy = locker.package.user_account.locker_policy

        free_date = locker.assignment_date + timedelta(days=policy.free_days_period)
        if datetime.now() <= free_date:
            return 0.0

        match locker.size.type:
            case LockerSizeType.SMALL:
                return 1
            case LockerSizeType.MEDIUM:
                return 2
            case LockerSizeType.LARGE:
                return 3

    def calculate_expiration_price(self, locker: Locker) -> float:
        price = self.calculate_price(locker)
        expiration_rate = 1.0

        return price * expiration_rate

File: src/go/oop/ShippingLocker/internal/price_calculator.go
package internal

import "time"

type PriceCalculator struct {
}

var (
	expirationRate float32 = 1.1
	smallRate      float32 = 1
	mediumRate     float32 = 2
	largeRate      float32 = 3
)

func (c *PriceCalculator) Calculate(locker *Locker) float32 {
	freeDate := locker.AssignDate.AddDate(0, 0, locker.pkg.User.Policy.FreeDaysPeriod)
	if time.Now().After(freeDate) {
		return 0
	}

	switch locker.Size.Type {
	case SmallLocker:
		return smallRate
	case MediumLocker:
		return mediumRate
	case LargeLocker:
		return largeRate
	}

	return mediumRate
}

func (c *PriceCalculator) CalculateExpirationPrice(locker *Locker) float32 {
	price := c.Calculate(locker)
	return price * expirationRate
}
